File: 20-actors/kabuto/methods/kotoba.py
#!/usr/bin/env python3
"""kotoba.py — kabuto kotoba Datom-log writer (local, content-addressed). ADR-2606022000
+ ADR-2605262130 + ADR-2605312345.

The substrate boundary (root CLAUDE.md): canonical state is the **kotoba Datom log** —
content-addressed EAVT assertions, append-only (非終末論). `methods/transact.py` is the OTHER write
path: an HTTP push into a *running* kotoba node (operator-gated). This module is the **local,
autonomous-loop** write path — the same path shionome / ipaddress / yabai / sukashi / watatsuna /
watari use (`methods/autorun.py`): a self-driving heartbeat appends content-addressed transactions
to a local append-only EDN log with NO external I/O, so kabuto can run its own
observe→analyze→persist public-company-supply-chain cycle on the Murakumo fleet without a human or
a live node in the loop.

Constitutional posture is preserved by construction (kabuto hard rules): every derived signal is a
RESILIENCE + accountability map (single-source / concentration / tier-depth / cross-bloc corridors)
— never a "who to hit" / raid / takeover target-list (G2); public listed-company public-record data
only (G1); concentration is an observation, never an antitrust/sanctions verdict (G4). The loop
persists exactly what `analyze.py` already computes, with derived signals flagged :supply/derived.

  - graph_datoms(rows)            → EAVT assertions for every entity (company / address / contact /
                                    supply-edge / process). E = the entity's id; lists fan out.
  - derived_datoms(companies, a)  → EAVT assertions for the analyzer's derived :supply/* signals,
                                    flagged :supply/derived (mirrors analyze.render_datoms).
  - make_tx / append_tx / read_log / head_cid / verify_chain — content-addressed commit-DAG.

EAVT = [op entity attribute value]; op is :db/add only (append-only — no :db/retract). Stdlib only.
Deterministic: the caller supplies tx_id + as_of (no wall clock) → resume-safe.
"""
from __future__ import annotations

import hashlib
import json
import os
import pathlib
import sys
from typing import Any

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from kabuto_edn import _parse, _tokens  # noqa: E402

LOG_DEFAULT = (pathlib.Path(__file__).resolve().parents[1] / "data"
               / "kabuto.datoms.kotoba.edn")

ID_KEYS = (":company/id", ":company.address/id", ":company.contact/id",
           ":supply.edge/id", ":company.process/id")


def _add(entity: str, attr: str, value: Any) -> list:
    """One append-only EAVT assertion: [:db/add <entity> <attr> <value>]."""
    return [":db/add", entity, attr, value]


def graph_datoms(rows: list) -> list[list]:
    """Flatten the public-company supply-chain graph into append-only EAVT assertions. E = the
    entity's id; cardinality-many list values fan out. Listed-company public-record facts only;
    the seed carries no personal PII (G1)."""
    out: list[list] = []
    for r in rows:
        if not isinstance(r, dict):
            continue
        e = next((r[k] for k in ID_KEYS if k in r), None)
        if e is None:
            continue
        for k, v in r.items():
            if k in ID_KEYS:
                continue
            for item in (v if isinstance(v, list) else [v]):
                out.append(_add(e, k, item))
    return out


def derived_datoms(a: dict) -> list[list]:
    """Flatten the analyzer's derived :supply/* concentration signals into EAVT assertions, each
    flagged :supply/derived true (a RESILIENCE + accountability map recomputed on read, never
    re-ingested as fact, never a target-list — G2/G4). Mirrors analyze.render_datoms.
    `a` is analyze.analyze()."""
    out: list[list] = []
    for i, (c, commodity, sup, crit) in enumerate(a["single_source"]):
        e = f"supply-single-{i}"
        out += [_add(e, ":supply/single-source-customer", c), _add(e, ":supply/commodity", commodity),
                _add(e, ":supply/sole-supplier", sup), _add(e, ":supply/criticality", crit),
                _add(e, ":supply/derived", True)]
    for country, load in sorted(a["jurisdiction_load"].items(), key=lambda kv: -kv[1]):
        e = f"supply-juris-{country}"
        out += [_add(e, ":supply/jurisdiction", country),
                _add(e, ":supply/jurisdiction-load", round(load, 2)), _add(e, ":supply/derived", True)]
    for sup, load in sorted(a["systemic"].items(), key=lambda kv: -kv[1]):
        e = f"supply-systemic-{sup}"
        out += [_add(e, ":supply/systemic-supplier", sup),
                _add(e, ":supply/out-degree", a["out_deg"].get(sup, 0)),
                _add(e, ":supply/outward-criticality", round(load, 2)), _add(e, ":supply/derived", True)]
    for c, secs, load, idx in a["diversification"]:
        e = f"supply-divers-{c}"
        out += [_add(e, ":supply/diversification-customer", c), _add(e, ":supply/supplier-sectors", secs),
                _add(e, ":supply/inbound-criticality", load), _add(e, ":supply/diversification-index", idx),
                _add(e, ":supply/derived", True)]
    for n, ind, outd, score in a["intermediaries"]:
        e = f"supply-inter-{n}"
        out += [_add(e, ":supply/intermediary", n), _add(e, ":supply/in-degree", ind),
                _add(e, ":supply/out-degree", outd), _add(e, ":supply/betweenness", score),
                _add(e, ":supply/derived", True)]
    for n, d in a["tier_depth"]:
        e = f"supply-tier-{n}"
        out += [_add(e, ":supply/tier-depth-node", n), _add(e, ":supply/tier-depth", d),
                _add(e, ":supply/derived", True)]
    for bloc, load in sorted(a["bloc_load"].items(), key=lambda kv: -kv[1]):
        e = f"supply-bloc-{bloc}"
        out += [_add(e, ":supply/region-bloc", bloc), _add(e, ":supply/bloc-load", round(load, 2)),
                _add(e, ":supply/derived", True)]
    for commodity, nsup, hhi in a["commodity_hhi"]:
        e = f"supply-commodity-{str(commodity).lstrip(':')}"
        out += [_add(e, ":supply/commodity", str(commodity).lstrip(":")),
                _add(e, ":supply/commodity-suppliers", nsup), _add(e, ":supply/commodity-hhi", hhi),
                _add(e, ":supply/derived", True)]
    for corridor, load in a["cross_corridors"]:
        e = f"supply-corridor-{corridor}"
        out += [_add(e, ":supply/cross-bloc-corridor", corridor),
                _add(e, ":supply/cross-bloc-load", round(load, 2)), _add(e, ":supply/derived", True)]
    for c, score, ss, secs, load, cross in a["resilience"]:
        e = f"supply-resil-{c}"
        out += [_add(e, ":supply/resilience-customer", c), _add(e, ":supply/resilience-score", score),
                _add(e, ":supply/single-source-count", ss), _add(e, ":supply/derived", True)]
    for sec, cap in a.get("sector_cap_rank", []):
        e = f"supply-capsec-{str(sec).lstrip(':')}"
        out += [_add(e, ":supply/cap-sector", str(sec).lstrip(":")),
                _add(e, ":supply/cap-sector-busd", round(cap, 1)), _add(e, ":supply/derived", True)]
    if a.get("cap_count"):
        e = "supply-cap-hhi"
        out += [_add(e, ":supply/cap-hhi", a["cap_hhi"]), _add(e, ":supply/cap-total-busd", a["total_cap"]),
                _add(e, ":supply/cap-covered", a["cap_count"]),
                _add(e, ":supply/cap-coverage-pct", a["cap_coverage"]), _add(e, ":supply/derived", True)]
    return out


def _canonical(datoms: list[list], prev_cid: str) -> bytes:
    return json.dumps({"prev": prev_cid, "datoms": datoms},
                      ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode("utf-8")


def tx_cid(datoms: list[list], prev_cid: str = "") -> str:
    """Content address = sha256 over (prev_cid, datoms) → a commit-DAG."""
    return "b" + hashlib.sha256(_canonical(datoms, prev_cid)).hexdigest()


def make_tx(datoms: list[list], *, tx_id: int, as_of: int, prev_cid: str = "") -> dict:
    return {
        ":tx/id": tx_id,
        ":tx/as-of": as_of,
        ":tx/prev": prev_cid,
        ":tx/cid": tx_cid(datoms, prev_cid),
        ":tx/count": len(datoms),
        ":tx/datoms": datoms,
    }


def _edn_val(v: Any) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return repr(v)
    if isinstance(v, str):
        return v if v.startswith(":") else json.dumps(v, ensure_ascii=False)
    if isinstance(v, list):
        return "[" + " ".join(_edn_val(x) for x in v) + "]"
    return json.dumps(str(v), ensure_ascii=False)


def _tx_to_edn(tx: dict) -> str:
    datoms = " ".join("[" + " ".join(_edn_val(x) for x in d) + "]" for d in tx[":tx/datoms"])
    return (f'{{:tx/id {tx[":tx/id"]} :tx/as-of {tx[":tx/as-of"]} '
            f':tx/prev {json.dumps(tx[":tx/prev"])} :tx/cid {json.dumps(tx[":tx/cid"])} '
            f':tx/count {tx[":tx/count"]} :tx/datoms [{datoms}]}}')


def append_tx(tx: dict, log_path: pathlib.Path = LOG_DEFAULT) -> str:
    """Append ONE transaction to the append-only log (never rewrites). Returns the tx CID."""
    log_path.parent.mkdir(parents=True, exist_ok=True)
    if not log_path.exists():
        log_path.write_text(";; kabuto kotoba Datom log — append-only EAVT transactions "
                            "(content-addressed DAG). Resilience + accountability map, never a "
                            "target-list. DO NOT hand-edit. ADR-2606022000.\n", encoding="utf-8")
    with log_path.open("a", encoding="utf-8") as fh:
        fh.write(_tx_to_edn(tx) + "\n")
    return tx[":tx/cid"]


def read_log(log_path: pathlib.Path = LOG_DEFAULT) -> list[dict]:
    """Read the log back as a list of transaction dicts (uses the shared kabuto_edn reader)."""
    if not log_path.exists():
        return []
    txs = []
    for line in log_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith(";"):
            continue
        txs.append(_parse(_tokens(line)))
    return txs


def head_cid(log_path: pathlib.Path = LOG_DEFAULT) -> str:
    txs = read_log(log_path)
    return txs[-1][":tx/cid"] if txs else ""


def verify_chain(log_path: pathlib.Path = LOG_DEFAULT) -> dict:
    """Recompute every CID from its datoms + prev; verify the DAG is intact. {ok, length, broken_at}."""
    txs = read_log(log_path)
    prev = ""
    for i, tx in enumerate(txs):
        expect = tx_cid(tx.get(":tx/datoms", []), prev)
        if tx.get(":tx/cid") != expect or tx.get(":tx/prev") != prev:
            return {"ok": False, "length": len(txs), "broken_at": i}
        prev = tx[":tx/cid"]
    return {"ok": True, "length": len(txs), "broken_at": -1}


if __name__ == "__main__":
    import pathlib as _pl
    from kabuto_edn import load_edn

    here = _pl.Path(__file__).resolve().parents[1]
    g = here / "data" / "companies.merged.kotoba.edn"
    if not g.exists():
        g = here / "data" / "seed-public-companies.kotoba.edn"
    datoms = graph_datoms(load_edn(g))
    tx = make_tx(datoms, tx_id=1, as_of=20260609, prev_cid="")
    print(f"# kabuto kotoba Datom tx — {tx[':tx/count']} EAVT assertions, cid={tx[':tx/cid'][:18]}…")

File: 20-actors/kabuto/methods/kotoba.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/kotoba.py — the local content-addressed Datom-log writer.
(ns kabuto.methods.kotoba
  "kotoba.clj — kabuto kotoba Datom-log writer (local, content-addressed).
  ADR-2606022000 + ADR-2605262130 + ADR-2605312345.

  Canonical state is the kotoba Datom log — content-addressed EAVT assertions, append-only. The
  local autonomous-loop write path (the shionome/watatsuna shape). Outputs are resilience +
  accountability (G2 — concentration / single-source / tier-depth / cross-bloc), never a
  raid/takeover/target framing; listed-company public-record facts only (G1); non-adjudicating (G4).

    graph-datoms   → EAVT for every entity (company/address/contact/supply-edge/process)
    derived-datoms → EAVT for derived :supply/* concentration signals (flagged :supply/derived)
    make-tx / append-tx / read-log / head-cid / verify-chain — content-addressed commit-DAG

  :db/add only (append-only). Deterministic: caller supplies tx-id + as-of; autorun sorts datoms
  into canonical order before hashing so the CID is reproducible regardless of set-iteration order."
  (:require [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn log-default []
  (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "kabuto.datoms.kotoba.edn")))

(def id-keys #{:company/id :company.address/id :company.contact/id :supply.edge/id :company.process/id})
(defn- add-datom [e a v] [:db/add e a v])

(defn graph-datoms [rows]
  (vec (mapcat
        (fn [r]
          (when (map? r)
            (when-let [e (some #(get r %) id-keys)]
              (for [[k v] r :when (not (id-keys k))
                    item (if (sequential? v) v [v])]
                (add-datom e k item)))))
        rows)))

(defn derived-datoms
  "Flatten the analyzer's derived :supply/* signals into EAVT assertions (flagged
  :supply/derived true). `a` is analyze/analyze. Mirrors analyze.render_datoms."
  [a]
  (let [r2 (fn [x] (/ (Math/round (* (double x) 100.0)) 100.0))
        cm (fn [c] (str/replace (str c) #"^:" ""))]
    (vec
     (concat
      (mapcat (fn [i [c commodity sup crit]]
                (let [e (str "supply-single-" i)]
                  [(add-datom e :supply/single-source-customer c)
                   (add-datom e :supply/commodity commodity)
                   (add-datom e :supply/sole-supplier sup)
                   (add-datom e :supply/criticality crit)
                   (add-datom e :supply/derived true)]))
              (range) (:single-source a))
      (mapcat (fn [[country load]]
                (let [e (str "supply-juris-" country)]
                  [(add-datom e :supply/jurisdiction country)
                   (add-datom e :supply/jurisdiction-load (r2 load))
                   (add-datom e :supply/derived true)]))
              (sort-by (comp - val) (:jurisdiction-load a)))
      (mapcat (fn [[sup load]]
                (let [e (str "supply-systemic-" sup)]
                  [(add-datom e :supply/systemic-supplier sup)
                   (add-datom e :supply/out-degree (get-in a [:out-deg sup] 0))
                   (add-datom e :supply/outward-criticality (r2 load))
                   (add-datom e :supply/derived true)]))
              (sort-by (comp - val) (:systemic a)))
      (mapcat (fn [[c secs load idx]]
                (let [e (str "supply-divers-" c)]
                  [(add-datom e :supply/diversification-customer c)
                   (add-datom e :supply/supplier-sectors secs)
                   (add-datom e :supply/inbound-criticality load)
                   (add-datom e :supply/diversification-index idx)
                   (add-datom e :supply/derived true)]))
              (:diversification a))
      (mapcat (fn [[n ind outd score]]
                (let [e (str "supply-inter-" n)]
                  [(add-datom e :supply/intermediary n)
                   (add-datom e :supply/in-degree ind)
                   (add-datom e :supply/out-degree outd)
                   (add-datom e :supply/betweenness score)
                   (add-datom e :supply/derived true)]))
              (:intermediaries a))
      (mapcat (fn [[n d]]
                (let [e (str "supply-tier-" n)]
                  [(add-datom e :supply/tier-depth-node n)
                   (add-datom e :supply/tier-depth d)
                   (add-datom e :supply/derived true)]))
              (:tier-depth a))
      (mapcat (fn [[bloc load]]
                (let [e (str "supply-bloc-" bloc)]
                  [(add-datom e :supply/region-bloc bloc)
                   (add-datom e :supply/bloc-load (r2 load))
                   (add-datom e :supply/derived true)]))
              (sort-by (comp - val) (:bloc-load a)))
      (mapcat (fn [[commodity nsup hhi]]
                (let [e (str "supply-commodity-" (cm commodity))]
                  [(add-datom e :supply/commodity (cm commodity))
                   (add-datom e :supply/commodity-suppliers nsup)
                   (add-datom e :supply/commodity-hhi hhi)
                   (add-datom e :supply/derived true)]))
              (:commodity-hhi a))
      (mapcat (fn [[corridor load]]
                (let [e (str "supply-corridor-" corridor)]
                  [(add-datom e :supply/cross-bloc-corridor corridor)
                   (add-datom e :supply/cross-bloc-load (r2 load))
                   (add-datom e :supply/derived true)]))
              (:cross-corridors a))
      (mapcat (fn [[c score ss _secs _load _cross]]
                (let [e (str "supply-resil-" c)]
                  [(add-datom e :supply/resilience-customer c)
                   (add-datom e :supply/resilience-score score)
                   (add-datom e :supply/single-source-count ss)
                   (add-datom e :supply/derived true)]))
              (:resilience a))))))

;; ── content-addressed commit-DAG ──────────────────────────────────────────────
(defn- sha256-hex [^String s]
  (let [md (java.security.MessageDigest/getInstance "SHA-256")]
    (apply str (map #(format "%02x" (bit-and % 0xff)) (.digest md (.getBytes s "UTF-8"))))))
(defn- canonical [datoms prev] (str "{:datoms " (pr-str datoms) " :prev " (pr-str prev) "}"))
(defn tx-cid
  ([datoms] (tx-cid datoms ""))
  ([datoms prev] (str "b" (sha256-hex (canonical datoms prev)))))

(defn canonical-order
  "Sort datoms into a canonical (pr-str) order so the CID is reproducible regardless of
  analyze's set-iteration order. Idempotent."
  [datoms]
  (vec (sort-by pr-str datoms)))

(defn make-tx [datoms & {:keys [tx-id as-of prev-cid] :or {prev-cid ""}}]
  {:tx/id tx-id :tx/as-of as-of :tx/prev prev-cid
   :tx/cid (tx-cid datoms prev-cid) :tx/count (count datoms) :tx/datoms datoms})

(defn append-tx
  ([tx] (append-tx tx (log-default)))
  ([tx log-path]
   (let [f (io/file log-path)]
     (.mkdirs (.getParentFile (.getAbsoluteFile f)))
     (when-not (.exists f)
       (spit f (str ";; kabuto kotoba Datom log — append-only EAVT transactions (content-addressed "
                    "DAG). Resilience + accountability, never a target-list (G2/G4). DO NOT hand-edit. "
                    "ADR-2606022000.\n")))
     (spit f (str (pr-str tx) "\n") :append true)
     (:tx/cid tx))))

(defn read-log
  ([] (read-log (log-default)))
  ([log-path]
   (let [f (io/file log-path)]
     (if-not (.exists f) []
       (->> (str/split-lines (slurp f)) (map str/trim)
            (remove #(or (empty? %) (str/starts-with? % ";"))) (mapv edn/read-string))))))

(defn head-cid
  ([] (head-cid (log-default)))
  ([log-path] (let [txs (read-log log-path)] (if (seq txs) (:tx/cid (last txs)) ""))))

(defn verify-chain
  ([] (verify-chain (log-default)))
  ([log-path]
   (let [txs (read-log log-path)]
     (loop [i 0 prev "" xs txs]
       (if (empty? xs)
         {:ok true :length (count txs) :broken-at -1}
         (let [tx (first xs) expect (tx-cid (:tx/datoms tx []) prev)]
           (if (or (not= (:tx/cid tx) expect) (not= (:tx/prev tx) prev))
             {:ok false :length (count txs) :broken-at i}
             (recur (inc i) (:tx/cid tx) (rest xs)))))))))
